File: C++/Step15-Stack/4949.cpp
//
//  4949.cpp
//  Step15-Stack
//
//  Created by 김예빈 on 2019. 9. 19..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <stack>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    string str;
    bool check;
    
    while(1) {
        getline(cin, str);
        stack<char> st;
        
        check = true;
        
        if(str[0] == '.' && str.length() == 1)
            break;
        
        for(int i = 0; i < str.length(); i++) {
            if(str[i] == '(')
                st.push(str[i]);
            
            if(str[i] == ')') {
                if (!st.empty() && st.top() == '(')
                    st.pop();
                else {
                    check = false;
                    break;
                }
            }
            
            if(str[i] == '[')
                st.push(str[i]);
            
            if(str[i] == ']') {
                if (!st.empty() && st.top() == '[')
                    st.pop();
                else {
                    check = false;
                    break;
                }
            }
        }
        
        if(check && st.empty())
            cout << "yes\n";
        else
            cout << "no\n";
    }
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step18-Stacks_1/4949.swift
//
//  4949.swift
//  Step17-Stacks
//
//  Created by Yebin Kim on 2019/12/27.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

extension Character {
    var pair: Character {
        if self == ")" {
            return "("
        } else {
            return "["
        }
    }
}

while true {
    var st = Array<Character>()
    let line = readLine()!
    
    if line == "." { break }
    
    for (_, e) in line.enumerated() {
        if e == "(" || e == "[" {
            st.append(e)
        } else if e == ")" || e == "]" {
            if st.isEmpty {
                st.append(e)
                break
            } else {
                if e.pair != st.last! {
                    break
                }
                
                st.removeLast()
            }
        }
    }
    
    if st.isEmpty {
        print("yes")
    } else {
        print("no")
    }
}

//public struct Stack<T> {
//    private var stack = Array<T>()
//
//    public mutating func push(_ element: T) {
//        stack.append(element)
//    }
//
//    public mutating func pop() {
//        stack.removeLast()
//    }
//
//    public var top: T? {
//        return stack.last
//    }
//
//    public var isEmpty: Bool {
//        return stack.isEmpty
//    }
//}
//
//extension Character {
//    var pair: Character {
//        if self == ")" {
//            return "("
//        } else {
//            return "["
//        }
//    }
//}
//
//while true {
//    var st = Stack<Character>()
//    let line = readLine()!
//
//    if line == "." { break }
//
//    for (_, e) in line.enumerated() {
//        if e == "(" || e == "[" {
//            st.push(e)
//        } else if e == ")" || e == "]" {
//            if st.isEmpty {
//                st.push(e)
//                break
//            } else {
//                if e.pair != st.top! {
//                    break
//                }
//
//                st.pop()
//            }
//        }
//    }
//
//    if st.isEmpty {
//        print("yes")
//    } else {
//        print("no")
//    }
//}
